[WIP] Reduce gas consumption for ERC20 transfers

smarx opened the pull request:
## Description

This rewrites `AllowanceTarget.exceuteCall()` and `TokenSpenderFeature._spendERC20Tokens()` in assembly to save 2798 gas per `_spendERC20Tokens()` call.

## Testing instructions

No test changes are necessary. `yarn test` tests the new, rewritten versions.

## Types of changes

* Bug fix (non-breaking change which fixes an issue)

Specifically, this is just gas optimization. No change in behavior is expected.

## Checklist:

-   [x] Prefix PR title with `[WIP]` if necessary.
-   [x] Add tests to cover changes as needed.
-   [x] Update documentation as needed.
-   [ ] Add new entries to the relevant CHANGELOG.jsons.

smarx committed:
Rewrite AllowanceTarget.executeCall() and
TokenSpenderFeature._spendERC20Tokens() in assembly to reduce gas
consumption. Total gas savings: 2798 gas per _spendERC20Tokens() call.

hysz commented on contracts/zero-ex/contracts/src/external/AllowanceTarget.sol:
```diff
@@ -46,10 +46,23 @@ contract AllowanceTarget is
         onlyAuthorized
         returns (bytes memory resultData)
```
I reckon we could shave off some more gas by putting in the fallback and doing the `onlyAuthorized` check in assembly. Here's [sample code](https://github.com/0xProject/0x-monorepo/blob/development/contracts/asset-proxy/contracts/src/ERC20Proxy.sol#L52) from the `ERC20Proxy`.

smarx replied:
I will give that a try. Thanks.

smarx replied:
This is done in the latest commit. It saved an additional 256 gas. Thanks!

hysz commented on contracts/zero-ex/contracts/src/external/AllowanceTarget.sol:
```diff
@@ -46,10 +46,23 @@ contract AllowanceTarget is
         onlyAuthorized
         returns (bytes memory resultData)
     {
-        bool success;
```
We can also remove the dependency on `LibRichErrorsV06`.

smarx committed:
More executeCall() gas golfing:

- migrate executeCall() to the fallback function
- replace onlyAuthorized with assembly equivalent

smarx committed:
More AllowanceTarget gas optimization

* Add a specific AllowanceTarget.transferFrom().
* Drop ABI encoding of return data.
* Do rich revert in assembly.
